### PR TITLE
Deduplicate shared variables in HDF5 saves

### LINKED ISSUE #18419: Model weights are not de-duplicated on save
When a model shares a single variable across many layers, `model.weights` is deduped. Only one weight will be returned no matter how many times it is referenced.

However when a model is saved, every layer that has a ref to variable will save their own copy. Weights are not deduped on save.

This can lead to a funny flow where things appear to be working, but the saved weights are actually duplicated on disk, and a single variable is assigned multiple times during load. This colab shows the issue ->
https://colab.research.google.com/gist/mattdangerw/2fabb09d53bd19cb6407d219e34155ab/weight-sharing-weirdness.ipynb

### PR DESCRIPTION
**Pull request by atharvaHJoshi**
## Description

Fixes #18419.

When a single variable is shared across layers, `model.weights` is deduplicated in memory, but saving still writes one copy per referencing layer. `model.save()` therefore stores every weight copy in the `.keras` file and `save_weights()` does the same for `.weights.h5`, doubling the artifact size and writing each shared variable multiple times on load.

This change makes the HDF5 stores (`H5IOStore` and `ShardedH5IOStore`) track which `Variable` objects have already been written. When the same variable is stored again, a single HDF5 hard link is created to the existing dataset instead of a second copy of its data.

Measured effect on a model where one 1M-float kernel is shared by two layers: `.weights.h5` size drops from ~8.0 MB to ~4.0 MB.

Sharded saves keep the metadata/weight-map accounting (and the size guard) based on genuinely-stored data only. If a duplicate variable would land in a different shard file (where an HDF5 hard link is not possible), it falls back to storing a plain copy rather than silently dropping data.

## Changes

- `keras/src/saving/saving_lib.py`:
  - `H5IOStore.__setitem__`: record written variables by object identity and hard-link repeats to the original dataset (`bfloat16` dtype attrs are preserved since the data is shared).
  - `ShardedH5IOStore.__setitem__`: skip size accounting when a repeat can be hard-linked within the current shard, with a cross-shard fallback to a plain copy.
- `keras/src/saving/saving_lib_test.py`: regression tests covering full model save, weights-only save, and sharded save with two layers sharing one kernel, asserting the two kernel datasets share the same on-disk address and that reloading reproduces the original outputs.

## Testing

- New regression tests pass on the JAX and NumPy backends.
- `keras/src/saving/saving_lib_test.py` full suite passes (the only failures are two pre-existing remote-path tests that require the TensorFlow `gfile` dependency, absent on this machine).
- `ruff check` and `ruff format --check` pass.

## Contributor Agreement

- [x] This PR is linked to an issue that has been assigned to me (see `Fixes #18419` above).
- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/saving/saving_lib.py [hunk 1]
```diff
@@ -1332,6 +1332,10 @@ def __init__(self, path_or_io, archive=None, mode="r"):
         self._h5_entry_metadata = None
         self._h5_entry_initialized = False
 
+        # Track already-written variables so that shared variables are stored
+        # as HDF5 hard links instead of duplicated datasets.
+        self._written_variable_datasets = {}
```
Comment (JyotinderSingh) [inline keras/src/saving/saving_lib.py:1337]: Keep the variable object itself, not only its `id()`, so the id cannot be recycled while the store is open (see the comment further down, on the line that records it).

```suggestion
        # Variables already written to this store, keyed by `id()`, so that
        # a shared variable is stored once and hard-linked afterwards. The
        # variable object is kept alive so its `id()` cannot be recycled.
        self._written_variables = {}
```

File: keras/src/saving/saving_lib.py [hunk 2]
```diff
@@ -1467,12 +1471,35 @@ def __setitem__(self, key, value):
...
+                # to the existing dataset instead of storing a second copy.
+                self._h5_entry_group[key] = self.h5_file[linked_dataset_path]
+                return
+            except KeyError:
+                # The previously written dataset is not reachable from the
+                # current H5 file (e.g. it lives in another shard). Fall
+                # through and store a new copy.
+                pass
+
         value = backend.convert_to_numpy(value)
```
Comment (JyotinderSingh) [inline keras/src/saving/saving_lib.py:1488]: #23642 renamed this to `backend.ops.convert_to_numpy` (same for the call in the sharded store below); that is the whole merge conflict.

File: keras/src/saving/saving_lib.py [hunk 3]
```diff
@@ -1467,12 +1471,35 @@ def __setitem__(self, key, value):
...
         if backend.standardize_dtype(value.dtype) == "bfloat16":
             ds = self._h5_entry_group.create_dataset(key, data=value)
             ds.attrs["dtype"] = "bfloat16"
         else:
             self._h5_entry_group[key] = value
+            ds = self._h5_entry_group[key]
+
+        if variable_id is not None:
+            self._written_variable_datasets[variable_id] = ds.name
```
Comment (JyotinderSingh) [inline keras/src/saving/saving_lib.py:1497]: This records `id(value)` without keeping `value` alive. If a `save_own_variables` writes a `Variable` that nothing else references, CPython recycles its `id`, and the next variable written is hard-linked to the previous dataset with no error:

```python
class Transient(keras.layers.Layer):
    def build(self, input_shape):
        self.w = self.add_weight(shape=(2,), initializer="zeros")

    def call(self, x):
        return x

    def save_own_variables(self, store):
        store["a"] = backend.Variable(np.array([1.0, 2.0], "float32"), trainable=False)
        store["b"] = backend.Variable(np.array([3.0, 4.0], "float32"), trainable=False)
```

On numpy, jax and torch the id is reused by the very next allocation, and `b` reads back as `[1., 2.]`, with both datasets at the same HDF5 address. No layer in the repo writes transient variables (the 11 `save_own_variables` overrides write tracked variables, plain tensors, or numpy arrays), but this is silent corruption for user code, and the fix is one line: store `(value, ds.name)` instead of `ds.name`. With that, the repro stores `b` correctly. The suggestion on this method does that.

File: keras/src/saving/saving_lib.py [hunk 4]
```diff
@@ -1467,12 +1471,35 @@ def __setitem__(self, key, value):
...
+
+    def _get_variable_id(self, value):
+        if isinstance(value, backend.Variable):
+            return id(value)
+        return None
```
Comment (JyotinderSingh) [inline keras/src/saving/saving_lib.py:1502]: Suggested shape for the writer: one method that takes an optional pre-converted array, so the sharded store can convert once (see below); the variable kept alive next to its path; and `dataset_path in self.h5_file` instead of `try`/`except KeyError` as control flow for "is this path in the current file" (it is what `has_path` already relies on). The sharded store currently probes the path twice per write, once for `is_linked` and again here; the helper folds both into one. `__setitem__` keeps the mode check above and only delegates.

```suggestion
        self._write(key, value)

    def _write(self, key, value, numpy_value=None):
        """Write `value` under `key` in the current H5 entry group.

        A `backend.Variable` that was already written to the current H5 file
        is stored as a hard link to the existing dataset instead of a copy.

        Args:
            key: `str`. The dataset name within the current H5 entry group.
            value: The value to write.
            numpy_value: Optional NumPy array holding the already-converted
                `value`, to avoid converting it twice.
        """
        if not self._h5_entry_initialized:
            self._create_h5_group(self._h5_entry_path)

        linked_dataset_path = self._get_linked_dataset_path(value)
        if linked_dataset_path is not None:
            self._h5_entry_group[key] = self.h5_file[linked_dataset_path]
            return

        if numpy_value is None:
            numpy_value = backend.convert_to_numpy(value)
        if backend.standardize_dtype(numpy_value.dtype) == "bfloat16":
            ds = self._h5_entry_group.create_dataset(key, data=numpy_value)
            ds.attrs["dtype"] = "bfloat16"
        else:
            self._h5_entry_group[key] = numpy_value
            ds = self._h5_entry_group[key]

        if isinstance(value, backend.Variable):
            self._written_variables[id(value)] = (value, ds.name)

    def _get_linked_dataset_path(self, value):
        """Return the path of the dataset already written for `value`.

        Returns `None` if `value` is not a `backend.Variable`, has not been
        written yet, or was written to a different H5 file (a previous
        shard), in which case it must be stored again.
        """
        if not isinstance(value, backend.Variable):
            return None
        written = self._written_variables.get(id(value))
        if written is None:
            return None
        _, dataset_path = written
        if dataset_path not in self.h5_file:
            return None
        return dataset_path
```

File: keras/src/saving/saving_lib.py [hunk 5]
```diff
@@ -1532,6 +1559,10 @@ def __init__(self, path_or_io, max_shard_size=5, archive=None, mode="r"):
         self._h5_entry_metadata = None
         self._h5_entry_initialized = False
 
+        # Track already-written variables so that shared variables are stored
+        # as HDF5 hard links instead of duplicated datasets.
+        self._written_variable_datasets = {}
```
Comment (JyotinderSingh) [inline keras/src/saving/saving_lib.py:1564]: Same as in `H5IOStore`.

```suggestion
        # Variables already written to this store, keyed by `id()`, so that
        # a shared variable is stored once and hard-linked afterwards. The
        # variable object is kept alive so its `id()` cannot be recycled.
        self._written_variables = {}
```

File: keras/src/saving/saving_lib.py [hunk 6]
```diff
@@ -1740,29 +1771,45 @@ def __getitem__(self, key):
...
+                    self.h5_file[linked_dataset_path]
+                    is_linked = True
+                except KeyError:
+                    # The previously written dataset is not reachable from the
+                    # current shard file. Fall through and store a new copy.
+                    pass
+
+        if not is_linked:
+            # Accumulate `current_shard_size`.
+            value_for_size = backend.convert_to_numpy(value)
```
Comment (JyotinderSingh) [inline keras/src/saving/saving_lib.py:1791]: Gemini is right about this one. This converts the value for size accounting, then `super().__setitem__(key, value)` converts the original variable again. Master converted once and passed the array down. Real (non-ndarray) `convert_to_numpy` calls for a three-layer `Dense` model with 6 variables:

| | unsharded | sharded |
|---|---|---|
| master | 6 | 6 |
| this PR | 6 | 12 |

Sharding exists for large models, and on GPU/TPU each call is a device-to-host copy, so this doubles the transfer volume of every sharded save. Convert once here and hand the array to the base writer (next two suggestions).

File: keras/src/saving/saving_lib.py [hunk 7]
```diff
@@ -1740,29 +1771,45 @@ def __getitem__(self, key):
...
+            dtype = backend.standardize_dtype(value_for_size.dtype)
+            weight_counts = math.prod(value_for_size.shape)
```
Comment (JyotinderSingh) [inline keras/src/saving/saving_lib.py:1793]: Converts once and reuses the array in the writer. The mode check moves up front since `_write` no longer performs it, and the second probe of the linked path goes away.

```suggestion
        if self.mode not in ("w", "a"):
            raise ValueError("Setting a value is only allowed in write mode.")
        self._restore_h5_file()

        numpy_value = None
        if self._get_linked_dataset_path(value) is None:
            # Accumulate `current_shard_size`. A hard-linked repeat of an
            # already-written variable takes up no space.
            numpy_value = backend.convert_to_numpy(value)
            dtype = backend.standardize_dtype(numpy_value.dtype)
            weight_counts = math.prod(numpy_value.shape)
```

File: keras/src/saving/saving_lib.py [hunk 8]
```diff
@@ -1740,29 +1771,45 @@ def __getitem__(self, key):
...
-            self.h5_file = self._create_new_shard_file()
-            self.current_shard_size = value_size
+            # Create a new shard if the current shard is full.
+            self.current_shard_size += value_size
+            if self.current_shard_size > self.max_shard_size:
+                self.close()
+                self.h5_file = self._create_new_shard_file()
+                self.current_shard_size = value_size
 
         super().__setitem__(key, value)
```
Comment (JyotinderSingh) [inline keras/src/saving/saving_lib.py:1814]:
```suggestion
        self._write(key, value, numpy_value)
```

File: keras/src/saving/saving_lib_test.py [hunk 9]
```diff
@@ -87,6 +87,34 @@ def load_own_variables(self, store):
...
+            shape=(1,),
+            name="bias",
+            initializer="zeros",
+        )
+
+    def call(self, inputs):
+        return ops.matmul(inputs, self.kernel) + self.bias
+
+    def get_config(self):
+        return super().get_config()
```
Comment (JyotinderSingh) [inline keras/src/saving/saving_lib_test.py:115]: This override exists only to satisfy the `__init__`-argument check, and it deliberately drops `shared_kernel`, which is why sharing cannot survive `load_model`. A one-line comment saying so would save the next reader a puzzle.

File: keras/src/saving/saving_lib_test.py [hunk 10]
```diff
@@ -1428,6 +1456,107 @@ def is_remote_path(path):
...
+
+    def test_sharded_save_weights_deduplicates_shared_variables(self):
+        model = self._get_shared_kernel_model()
+        temp_filepath = os.path.join(
+            self.get_temp_dir(), "shared_variable_model.weights.json"
+        )
+        x = np.array([[1.0, 2.0]])
+        ref_output = model(x)
+
+        model.save_weights(temp_filepath, max_shard_size=0.005)
```
Comment (JyotinderSingh) [inline keras/src/saving/saving_lib_test.py:1538]: The current test does not exercise the cross-shard fallback. With a 2×1 kernel and `max_shard_size=0.005`, everything is stored in one shard, so the `except KeyError` branches flagged by Codecov never run.

Please add a test that forces the shared variable across a shard boundary. One setup that works is:

* Two layers share a 32×32 `float32` kernel (4 KB).
* The second layer creates another 32×32 weight before tracking the shared kernel.
* `max_shard_size=6e-6`.

This produces three shards. The shared kernel is copied into the third shard, `total_size` is 12,288 bytes, the second layer maps to two shard files, and the reloaded model produces matching results.

One important detail: assign the shared variable in `build`, after calling `add_weight` for the layer’s own variable. If it is assigned in `__init__`, attribute tracking places it first, so it never crosses a shard boundary.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Before this moves forward, please rebase the branch and address the double conversion in the sharded store. There is also one hardening change to make and one missing test for the fallback path. I’ve left the details inline, including committable suggestions for the store changes.

Some context, since this is the second attempt at #18419: the store logic is the same as in #23174, which was closed as stale, except that this version omits the `_warn_if_duplicate_own_variables` warning that @hertschuh questioned there. Please mention this history in the PR description.

Regarding that question: nested layers do not re-report their children’s variables. `Layer.save_own_variables` writes only `_trainable_variables + _non_trainable_variables`. For example, a Sequential model nested inside a Functional model, where the same `Dense` instance is applied twice, produces no hard links: 26 datasets map to 26 distinct HDF5 objects, including optimizer state. Deduplication occurs only when the same `Variable` object is tracked by two different layer instances, which is the scenario described in the issue.

The description should also clarify the scope and behavior of this change:

* It changes only what `save` and `save_weights` write.
* `load_model` still reconstructs layers from their configs, so the reloaded layers own independent kernels: `la.kernel is lb.kernel` is `False`.
* A weights-only load still assigns the shared variable once for each referencing layer, resulting in two assignments.
* Model outputs match in both cases.
* Existing readers remain compatible: Keras 3.11.3 can load the `.weights.h5`, `.keras`, and sharded files produced by this branch.

There are also two known limitations, neither of which requires action in this PR:

* A variable referenced from multiple shards is stored once in each shard.
* `KerasFileEditor.save()` rewrites links as ordinary copies because the editor operates on arrays rather than variables.

The inline suggestions together form a single patch. With all of them applied:

* The three new tests pass on NumPy, JAX, and Torch.
* The complete `saving_lib_test.py` suite passes on JAX.
* Sharded saves perform one conversion per variable again.
* The object-ID reuse reproduction stores both values correctly.
* The cross-shard fallback continues to copy as intended.

The suggestions are written against the current branch. After rebasing, the two calls to `backend.convert_to_numpy` should become `backend.ops.convert_to_numpy`.
